quipucords/scanner/network/utils.py

```diff
@@ -173,3 +181,54 @@ def get_fact_names():
 def raw_facts_template():
     """Results template for fact collection on network scans."""
     return {fact_name: None for fact_name in get_fact_names()}
+
+
+GEN_KEYFILES_DIRECTORY_PREFIX = "quipucords_ssh_private_key_files_"
+GEN_KEYFILE_PREFIX = "ssh_private_key_file_credential_"
+
+
+def generate_ssh_keyfile(credential, ssh_keyvalue):
+    """Generate an ssh private keyfile for the ssh key content specified."""
+    private_keyfile_path = os.path.join(
+        tempfile.mkdtemp(prefix=GEN_KEYFILES_DIRECTORY_PREFIX),
+        f"{GEN_KEYFILE_PREFIX}{credential.get('id')}",
+    )
+    private_key_fp = open(private_keyfile_path, "w+")
+    private_key_fp.write(ssh_keyvalue)
+    private_key_fp.write("\n")
+    private_key_fp.close()
+    os.chmod(private_keyfile_path, 0o600)
+    return private_keyfile_path
+
+
+def delete_ssh_keyfiles(inventory):
+    """Delete all generated ansible_ssh_private_key_files."""
+    if not inventory:
+        return
+
+    # Defined during connects
+    ansible_vars = inventory.get("all", {}).get("vars", {})
+    ssh_pkf = ansible_vars.get("ansible_ssh_private_key_file")
+    if ssh_pkf:
+        delete_ssh_keyfile(ssh_pkf)
+    # Defined during inspects
+    children = inventory.get("all", {}).get("children", {})
+    for group in children.keys():
+        hosts = children.get(group).get("hosts", {})
+        for host in hosts.keys():
+            ssh_pkf = hosts.get(host).get("ansible_ssh_private_key_file")
+            if ssh_pkf:
+                delete_ssh_keyfile(ssh_pkf)
+
+
+def delete_ssh_keyfile(private_keyfile_path):
+    """Remove a generated ssh_keyfile and delete its empty parent directory."""
+    keyfile_pattern = (
+        f"^/.*/{GEN_KEYFILES_DIRECTORY_PREFIX}.*/{GEN_KEYFILE_PREFIX}[0-9]+$"
+    )
+    if re.match(keyfile_pattern, private_keyfile_path):
+        if os.path.isfile(private_keyfile_path):
+            os.remove(private_keyfile_path)
+            parent_dir = os.path.abspath(os.path.join(private_keyfile_path, os.pardir))
+            if not os.listdir(parent_dir):
+                os.rmdir(parent_dir)
```

Comment:
I'm pretty sure this will break things unexpectedly in Celery workers when performing more than one scan, maybe also breaking the legacy threaded/multiprocessing manager implementation. Imagine two threads or processes on the same machine accessing keyfiles, and one deletes before the other has completed its work. 💥 

I believe we need to use something like [tempfile.TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) and keep an in-memory reference to the directory so we can clean it up (assuming again that we can't convert this whole file handling process into a context manager, but I don't think that's happening any time soon).

Reply:
From what it looks like, tempfile.TemporaryDirectory === tempfile.mkdtemp, so I think we're safe. But, yea going to random filenames would also protect this case.

Reply:
Sorry for crossing the streams!  😆  See the _other_ comment thread with a local demo comparing `tempfile.TemporaryDirectory` against `tempfile.mkdtemp`. `TemporaryDirectory` gives an automatic cleanup on process shutdown as a nice bonus.

Reply:
Thank you for verifying 👍  from what I've read, except for the cleanup done with TemporaryDirectory, they are identical.